### PR TITLE
Add upstream proxy support

### PR DESCRIPTION
**Pull request by jim3ma**
Proxy chains is useful for some scenario when the network is protected by some gateway.

This PR adds a option named `upstream` like:

```
forwardproxy {
    upstream https://upstearm.proxy.site
}
```

The upstream proxy supports multiple types proxy, like `https://a.com`, `http://b.com`, `sock5://c.com`, `http2://d.com`, for this we import a sub library `proxy` in `github.com/yaproxy/libyap`

And I don't know how caddy deals with dependency.
The official tool `dep` is good enough to do this.

**TODO**:

* multi upstream support ?
* customize DNS resolver ?

### CODE REVIEW

File: forwardproxy.go [hunk 1]
```diff
@@ -45,6 +45,8 @@ type ForwardProxy struct {
 	dialTimeout        time.Duration // for initial tcp connection
 	hostname           string        // do not intercept requests to the hostname (except for hidden link)
 	port               string        // port on which chain with forwardproxy is listening on
+	upstream           string
+	dial               func(network, address string) (net.Conn, error)
```
Comment (jim3ma) [inline forwardproxy.go]: `upstream` store the upstream server addr 
`dial` is a dial function for upstream proxy

File: setup.go [hunk 2]
```diff
@@ -165,11 +173,45 @@ func setup(c *caddy.Controller) error {
...
+		}).DialContext
+	}
+
+	if fp.dial == nil {
+		fp.dial = (&net.Dialer{
+			Timeout:   fp.dialTimeout,
+			KeepAlive: 30 * time.Second,
+			DualStack: true,
+		}).Dial
+	}
```
Comment (jim3ma) [inline setup.go]: add a default dial function with fp.dialTimeout

**Comment by mholt** [comment]
Thanks for the revised PR! One of us will get to this before too long.

**Comment by mholt** [comment]
Okay, having looked at this a little bit, what I think we need to do first is determine whether #20 fixes #17. If so, (or if it's close), that would definitely be the preferred way of handling it. Could you do us a favor and try #20 in conjunction with the `proxy` directive and report back?

**Comment by sergeyfrolov** [comment]
Thanks for the PR!
I very much like the idea of overriding the dialer to accomplish this, but we do not take new external dependencies lightly.
How about we override the dial with a function that simply always dials the upstream server, regardless of the actual destination?

File: forwardproxy.go (hunk 1 above)
Comment (sergeyfrolov) [reply forwardproxy.go]: Those are good comments, and they belong in code for future readers to see.

File: setup.go [hunk 3]
```diff
@@ -165,11 +173,45 @@ func setup(c *caddy.Controller) error {
 		}
 	}
 
-	fp.httpTransport.DialContext = (&net.Dialer{
-		Timeout:   fp.dialTimeout,
-		KeepAlive: 30 * time.Second,
-		DualStack: true,
-	}).DialContext
+	if fp.upstream != "" {
```
Comment (sergeyfrolov) [inline setup.go]: We can move this block up to where `case "upstream":` is

File: setup.go [hunk 4]
```diff
@@ -165,11 +173,45 @@ func setup(c *caddy.Controller) error {
...
+		fixedURL, err := url.Parse(fp.upstream)
+		if err != nil {
+			return c.ArgErr()
```
Comment (sergeyfrolov) [inline setup.go]: Better to return an actual error from parsing

File: setup.go [hunk 5]
```diff
@@ -165,11 +173,45 @@ func setup(c *caddy.Controller) error {
...
+			newDialer, err := proxy.FromURL(fixedURL, dialer, nil)
+			if err != nil {
+				return c.ArgErr()
+			}
+			fp.dial = newDialer.Dial
+			fp.httpTransport.Dial = newDialer.Dial
+		}
+	}
+
+	if fp.httpTransport.Dial == nil {
```
Comment (sergeyfrolov) [inline setup.go]: Could be in `else` block, corresponding to `if fp.upstream != ""`, making code easier to follow

File: setup.go [hunk 6]
```diff
@@ -165,11 +173,45 @@ func setup(c *caddy.Controller) error {
...
+		fp.httpTransport.DialContext = (&net.Dialer{
+			Timeout:   fp.dialTimeout,
+			KeepAlive: 30 * time.Second,
+			DualStack: true,
+		}).DialContext
+	}
+
+	if fp.dial == nil {
```
Comment (sergeyfrolov) [inline setup.go]: Could be in `else` block, corresponding to `if fp.upstream != ""`, making code easier to follow

**Comment by jim3ma** [comment]
The sub library just implements `RegisterDialerType` https://github.com/golang/net/tree/master/proxy which created by golang, and provides more upstream types such as socks5, https.

**Comment by sergeyfrolov** [comment]
@jim3ma sorry for not responding, had a personal emergency and forgot about this.
To repeat my previous comment, we do not take external dependencies lightly. I've discussed this matter with @mholt some time ago and we both are strongly against adding external dependencies.
Standard library, maintained by the Go authors, would be acceptable, however, it would be best not to add any new dependencies at all.
We are holding PRs to a high standards only to keep forwardproxy small, well-tested, stable and secure, so let's work together and find a way to make this PR mergeable.

**Comment by jeddytier4** [comment]
@jim3ma Might I suggest using https://godoc.org/golang.org/x/net/proxy instead of the one that you are using.  I believe that it will give the same results.

**Comment by sergeyfrolov** [comment]
@jeddytier4, `x/net/proxy` package looks great, thanks for the suggestion!
Packages under `golang.org/x` might be developed under looser compatibility requirements and be less stable in general, than the Go core, but they are part of Go Project.

**Comment by jeddytier4** [comment]
I will fork this and rewrite it based on x/net/proxy shortly, but have another pull request in review for this and don't want to break that.

**Comment by jim3ma** [comment]
Okay, I will replace the proxy library later.

**Comment by jim3ma** [comment]
@sergeyfrolov golang/x/net/proxy did not implement https proxy, we must do it ourselves by calling `golang/x/net/proxy.RegisterDialerType`

File: setup.go [hunk 7]
```diff
@@ -165,11 +173,41 @@ func setup(c *caddy.Controller) error {
...
-		DualStack: true,
-	}).DialContext
+	if fp.upstream != "" {
+		fixedURL, err := url.Parse(fp.upstream)
+		if err != nil {
+			return errors.New("wrong upstream address")
+		}
+		switch fixedURL.Scheme {
+		case "http":
+			fp.httpTransport.Proxy = http.ProxyURL(fixedURL)
```
Comment (sergeyfrolov) [inline setup.go]: Why do we set this up only for http://upstream.com?
`fp.httpTransport` is used when forwardproxy gets http request for the actual destination, and does not seem to be related to upstream scheme.

**Comment by sergeyfrolov** [comment]
Development of this feature continues in #27
